### PR TITLE
Restart applets if required

### PR DESCRIPTION
**Pull request by charlesbaynham**
# ARTIQ Pull Request

Restart existing applets if their spec changes

## Description of Changes

Add a call to `applet_dock.restart()` when applets' specs are updated via a ccb while they're already running.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] (Don't think this is noteworthy enough) Update [RELEASE_NOTES.md](../RELEASE_NOTES.md) if there are noteworthy changes, especially if there are changes to existing APIs.
- [x] Close/update issues.
- [ ] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments
- [ ] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

**Comment by sbourdeauducq** [comment]
>     * (Don't think this is noteworthy enough)

It definitely is - it's a clear change of behavior from an established API.

File: artiq/dashboard/applets_ccb.py [hunk 1]
```diff
@@ -181,9 +183,17 @@ def ccb_create_applet(self, name, command, group=None, code=None):
         else:
             spec = {"ty": "code", "code": code, "command": command}
         if applet is None:
+            logger.info('Applet {} does not exist: creating'.format(name))
```
Comment (sbourdeauducq) [inline artiq/dashboard/applets_ccb.py:186]: Should be ``debug`` imo.

**Comment by charlesbaynham** [comment]
> It definitely is - it's a clear change of behavior from an established API.

Fair enough. Would this go under ARTIQ-6, or are new features still being entered against ARTIQ-5?

**Comment by sbourdeauducq** [comment]
It's for ARTIQ-6.
Also I wonder what happens if applets are restarted too fast, in a way that could create several concurrent instances of the ``applet_dock.restart`` task. This problem isn't specific to your PR and it can also appear - in theory - with the GUI, but it is easier to expose it with code, which can run faster than one can click. Could you take a look?

File: artiq/dashboard/applets_ccb.py [hunk 2]
```diff
@@ -181,9 +183,17 @@ def ccb_create_applet(self, name, command, group=None, code=None):
         else:
             spec = {"ty": "code", "code": code, "command": command}
         if applet is None:
+            logger.debug('Applet {} does not exist: creating'.format(name))
```
Comment (sbourdeauducq) [inline artiq/dashboard/applets_ccb.py:186]: Oh and another thing: use formatting in logger and not ``.format()``. This avoids consuming resources when the log message isn't printed and it is a general coding pattern when using the logger. Yes, it only supports old-style format strings, but this is a problem with Python. This is the last thing AFAICT and then we can merge.
Also (but I have more or less given up enforcing that) ARTIQ generally uses " as string delimiter.

**Comment by charlesbaynham** [comment]
> Also I wonder what happens if applets are restarted too fast, in a way that could create several concurrent instances of the `applet_dock.restart` task. This problem isn't specific to your PR and it can also appear - in theory - with the GUI, but it is easier to expose it with code, which can run faster than one can click. Could you take a look?

Had a try with 

``` python
from artiq.experiment import EnvExperiment, NumberValue


default_cmd = "${{artiq_applet}}plot_xy {y_dataset}"
dataset_name = "test_data"


class TestCCB(EnvExperiment):

    def build(self):
        self.setattr_device('ccb')
        self.setattr_argument('repeats',
                              NumberValue(default=1, ndecimals=0, step=1))

    def run(self):
        self.set_dataset(dataset_name, [1, 2, 3, 2, 1], broadcast=True)

        titles = [
            "hello world {}".format(x) for x in range(0, self.repeats)
        ]

        for title in titles:
            cmd = (default_cmd + ' --title "{title}"').format(title=title, y_dataset=dataset_name)
            self.ccb.issue("create_applet",
                           "Test plot",
                           cmd,
                           )
```

The running applet seems to crash with

```
quamash.QEventLoop:Task exception was never retrieved
future: <Task finished coro=<AsyncioParentComm._autoclose() done, defined at C:\Users\cb6\.conda\envs\artiq-5\lib\site-packages\sipyco\pipe_ipc.py:118> exception=OSError(9, 'The handle is invalid', None, 6, None)>
Traceback (most recent call last):
  File "C:\Users\cb6\.conda\envs\artiq-5\lib\asyncio\tasks.py", line 240, in _step
    result = coro.send(None)
  File "C:\Users\cb6\.conda\envs\artiq-5\lib\site-packages\sipyco\pipe_ipc.py", line 120, in _autoclose
    self.server[0].close()
  File "C:\Users\cb6\.conda\envs\artiq-5\lib\asyncio\windows_events.py", line 283, in close
    pipe.close()
  File "C:\Users\cb6\.conda\envs\artiq-5\lib\asyncio\windows_utils.py", line 157, in close
    CloseHandle(self._handle)
OSError: [WinError 6] The handle is invalid
```

And then get relaunched. Even with repeats=1000 didn't cause any delay. I haven't dug into the source of the error message: it's not the best UX, but it's quite a contrived circumstance to be quickly altering the command of an applet, and the actual behaviour is ok.

**Comment by dnadlinger** [comment]
+1, we have been using something like this for ages internally (although implemented as literally unchecking/checking the tickbox).

**Pull request merged**